test/testframe.c

```diff
@@ -11,368 +11,504 @@
+        unsigned    newAlloc = MAX(1, TestAlloc * 2);
 
-        /* Reallocate array */
-        if (NULL == (newTest = (TestStruct *)realloc(Test, newAlloc * sizeof(TestStruct)))) {
-            printf("Out of memory for tests, Index = %u, TestAlloc = %u, newAlloc = %u\n", Index, TestAlloc,
-                   newAlloc);
-            exit(EXIT_FAILURE);
+        if (NULL == (newTest = realloc(TestArray, newAlloc * sizeof(TestStruct)))) {
+            fprintf(stderr,
+                    "%s: couldn't reallocate test array, TestCount = %u, TestAlloc = %u, newAlloc = %u\n",
+                    __func__, TestCount, TestAlloc, newAlloc);
+            return FAIL;
         }
 
-        /* Update info */
-        Test      = newTest;
+        TestArray = newTest;
         TestAlloc = newAlloc;
     }
 
-    /* Set up test function */
-    strcpy(Test[Index].Description, TheDescr);
-    if (*TheName != '-') {
-        strcpy(Test[Index].Name, TheName);
-        Test[Index].SkipFlag = 0;
+    /* If the test name begins with '-', skip the test by default */
+    if (*TestName == '-') {
+        TestArray[TestCount].TestSkipFlag = 1;
+        TestName++;
     }
-    else { /* skip test by default */
-        strcpy(Test[Index].Name, TheName + 1);
-        Test[Index].SkipFlag = 1;
+    else
+        TestArray[TestCount].TestSkipFlag = 0;
+
+    strcpy(TestArray[TestCount].Name, TestName);
+    strcpy(TestArray[TestCount].Description, TestDescr);
+
+    /* Make a copy of the additional test data given */
+    if (TestData) {
+        if (NULL == (new_test_data = malloc(TestDataSize))) {
+            fprintf(stderr, "%s: couldn't allocate space for additional test data\n", __func__);
+            return FAIL;
+        }
+
+        memcpy(new_test_data, TestData, TestDataSize);
     }
-    Test[Index].Call       = TheCall;
-    Test[Index].Cleanup    = Cleanup;
-    Test[Index].NumErrors  = -1;
-    Test[Index].Parameters = Parameters;
+    TestArray[TestCount].TestParameters = new_test_data;
 
-    /* Increment test count */
-    Index++;
+    TestArray[TestCount].TestFunc        = TestFunc;
+    TestArray[TestCount].TestSetupFunc   = TestSetupFunc;
+    TestArray[TestCount].TestCleanupFunc = TestCleanupFunc;
+    TestArray[TestCount].TestNumErrors   = -1;
+
+    TestCount++;
+
+    return SUCCEED;
 }
 
 /*
- * Initialize testing framework
- *
- * ProgName: Name of test program.
- * private_usage: Optional routine provided by test program to print the
- *      private portion of usage page.  Default to NULL which means none is
- *      provided.
- * private_parser: Optional routine provided by test program to parse the
- *      private options.  Default to NULL which means none is provided.
- *
+ * Initialize the testing framework
  */
-void
-TestInit(const char *ProgName, void (*private_usage)(void), int (*private_parser)(int ac, char *av[]))
+herr_t
+TestInit(const char *ProgName, void (*TestPrivateUsage)(FILE *stream),
```

Comment:
Added `FILE *` parameters here and to other functions that print output. Also added an extra `TestProcessID` parameter that is used to control output for parallel tests, while not requiring any MPI datatypes or calls in this function.